validator_worker/src/leader.rs

```diff
@@ -1,12 +1,37 @@
-use futures::future::FutureExt;
-use primitives::validator::{Validator, ValidatorFuture};
-use primitives::Channel;
+use std::error::Error;
 
-#[derive(Clone)]
-pub struct Leader {}
+use primitives::adapter::Adapter;
+use primitives::validator::{Accounting, MessageTypes, NewState};
+use primitives::BalancesMap;
 
-impl Validator for Leader {
-    fn tick(&self, _channel: Channel) -> ValidatorFuture<()> {
-        futures::future::ok(()).boxed()
+use crate::heartbeat::heartbeat;
+use crate::sentry_interface::SentryApi;
+use crate::{get_state_root_hash, producer};
+
+pub async fn tick<A: Adapter + 'static>(iface: &SentryApi<A>) -> Result<(), Box<dyn Error>> {
+    let (balances, new_accounting) = await!(producer::tick(&iface))?;
+
+    if let Some(new_accounting) = new_accounting {
+        on_new_accounting(&iface, (&balances, &new_accounting))?;
     }
+
+    await!(heartbeat(&iface, balances)).map(|_| ())
+}
+
+fn on_new_accounting<A: Adapter + 'static>(
+    iface: &SentryApi<A>,
+    (balances, new_accounting): (&BalancesMap, &Accounting),
+) -> Result<(), Box<dyn Error>> {
+    let state_root_raw = get_state_root_hash(&iface, &balances)?;
+    let state_root = hex::encode(state_root_raw);
+
+    let signature = iface.adapter.sign(&hex::encode(&state_root))?;
```

Comment:
you have double hex::encode() here, should be one

Reply:
My bad, thanks!

Reply:
Fixed.